br/pkg/stream/table_mapping.go

```diff
@@ -180,11 +137,148 @@ func (tc *TableMappingManager) parseTableValueAndUpdateIdMapping(dbID int64, val
 	return nil
 }
 
+func (tm *TableMappingManager) MergeBaseDBReplace(baseMap map[UpstreamID]*DBReplace) {
+	// merge baseMap to DBReplaceMap
+	for upstreamID, baseDBReplace := range baseMap {
+		if existingDBReplace, exists := tm.DBReplaceMap[upstreamID]; exists {
+			existingDBReplace.DbID = baseDBReplace.DbID
+
+			for tableUpID, baseTableReplace := range baseDBReplace.TableMap {
+				if existingTableReplace, tableExists := existingDBReplace.TableMap[tableUpID]; tableExists {
+					existingTableReplace.TableID = baseTableReplace.TableID
+
+					for partUpID, basePartDownID := range baseTableReplace.PartitionMap {
+						existingTableReplace.PartitionMap[partUpID] = basePartDownID
+					}
+				} else {
+					existingDBReplace.TableMap[tableUpID] = baseTableReplace
+				}
+			}
+		} else {
+			tm.DBReplaceMap[upstreamID] = baseDBReplace
+		}
+	}
+}
+
+func (tm *TableMappingManager) IsEmpty() bool {
+	return len(tm.DBReplaceMap) == 0
+}
+
+func (tm *TableMappingManager) ReplaceTemporaryIDs(
+	ctx context.Context, genGenGlobalIDs func(ctx context.Context, n int) ([]int64, error)) error {
+	if tm.tempIDCounter == InitialTempId {
+		// no temporary IDs were allocated
+		return nil
+	}
+
+	// find actually used temporary IDs
+	usedTempIDs := make(map[DownstreamID]struct{})
+
+	// Helper function to check and add temporary ID
+	addTempIDIfNeeded := func(id DownstreamID) error {
+		if id < 0 {
+			if _, exists := usedTempIDs[id]; exists {
+				return errors.Annotate(berrors.ErrRestoreInvalidRewrite,
+					fmt.Sprintf("found duplicate temporary ID: %d", id))
+			}
+			usedTempIDs[id] = struct{}{}
+		}
+		return nil
+	}
+
+	// check DBReplaceMap for used temporary IDs
+	// any value less than 0 is temporary ID
+	for _, dr := range tm.DBReplaceMap {
+		if err := addTempIDIfNeeded(dr.DbID); err != nil {
+			return err
+		}
+		for _, tr := range dr.TableMap {
+			if err := addTempIDIfNeeded(tr.TableID); err != nil {
+				return err
+			}
+			for _, partID := range tr.PartitionMap {
+				if err := addTempIDIfNeeded(partID); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	tempIDs := make([]DownstreamID, 0, len(usedTempIDs))
+	// convert to sorted slice
+	for id := range usedTempIDs {
+		tempIDs = append(tempIDs, id)
+	}
+
+	// sort to -1, -2, -4, -8 ... etc
```

Comment:
`tempIDs[i] > tempIDs[j]` does not means `upstream(tempIDs[i]) > upstream[tempIDs[j])`, so is it necessary to sort tempIDs?

Reply:
yes it doesn't, I was thinking to assign id based on first seen first, somewhat like sorted by time in the log backup, I was hoping it can be a bit useful when debugging. The order doesn't suggest anything such as upstream id ordering.